src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs

```diff
@@ -69,49 +86,64 @@ public ProjectImportsCollector(string logFilePath, bool createFile, string sourc
                 _fileStream = null;
                 _zipArchive = null;
             }
+            _runOnBackground = runOnBackground;
         }
 
-        public void AddFile(string filePath)
+        public void AddFile(string? filePath)
+        {
+            AddFileHelper(filePath, AddFileCore);
+        }
+
+        public void AddFileFromMemory(
+            string? filePath,
+            string data,
+            DateTimeOffset? entryCreationStamp = null,
+            bool makePathAbsolute = true)
+        {
+            AddFileHelper(filePath, path =>
+                AddFileFromMemoryCore(path, data, makePathAbsolute, entryCreationStamp));
+        }
+
+        public void AddFileFromMemory(
+            string? filePath,
+            Stream data,
+            DateTimeOffset? entryCreationStamp = null,
+            bool makePathAbsolute = true)
+        {
+            AddFileHelper(filePath, path => AddFileFromMemoryCore(path, data, makePathAbsolute, entryCreationStamp));
+        }
+
+        private void AddFileHelper(
+            string? filePath,
+            Action<string> addFileWorker)
         {
             if (filePath != null && _fileStream != null)
             {
                 lock (_fileStream)
                 {
-                    // enqueue the task to add a file and return quickly
-                    // to avoid holding up the current thread
-                    _currentTask = _currentTask.ContinueWith(t =>
+                    if (_runOnBackground)
+                    {
+                        // enqueue the task to add a file and return quickly
+                        // to avoid holding up the current thread
+                        _currentTask = _currentTask.ContinueWith(
+                            t => { TryAddFile(); },
+                            TaskScheduler.Default);
+                    }
+                    else
                     {
-                        try
-                        {
-                            AddFileCore(filePath);
-                        }
-                        catch
-                        {
-                        }
-                    }, TaskScheduler.Default);
+                        TryAddFile();
+                    }
                 }
             }
-        }
 
-        public void AddFileFromMemory(string filePath, string data)
-        {
-            if (filePath != null && data != null && _fileStream != null)
+            void TryAddFile()
```

Comment:
>void [](http://example.com/codeflow?start=12&length=4)

The name here indicates a 'Try' pattern but doesn't return a boolean indicating successful add or not.

Reply:
fixed